designs/2018-programmatic-style-edits/README.md

```diff
@@ -0,0 +1,401 @@
+        n.FunctionDeclaration.assert(add);
+
+        // If you choose to use recast.builders to construct new AST nodes, all builder
+        // arguments will be dynamically type-checked against the Mozilla Parser API.
+        var b = recast.types.builders;
+
+        // This kind of manipulation should seem familiar if you've used Esprima or the
+        // Mozilla Parser API before.
+        ast.program.body[0] = b.variableDeclaration("var", [
+            b.variableDeclarator(add.id, b.functionExpression(
+                null, // Anonymize the function expression.
+                add.params,
+                add.body
+            ))
+        ]);
+
+        // Just for fun, because addition is commutative:
+        add.params.push(add.params.shift());
+
+        // return the modified code
+        return recast.print(ast).code;
+    }
+}
+```
+
+Any JavaScript utility that can work on a string and return a string can be used inside a `text` style edit.
+
+Because the style edits have a defined order inside of a style editor, end-users can prescribe any number of steps to edit styles in source code. For example, you might want to run a Recast edit and then a Prettier edit:
+
+```js
+// style editor
+module.exports = {
+    meta: {
+        name: "ES5 Style",
+        description: "A style editor just for ES5"
+    },
+    edits: [
+        require("./edits/recast"),
+        require("./edits/prettier")
+    ]
+}
+```
+
+This ability to order individual edits means that end-users could potentially run Prettier to get a baseline format for their source code and then run additional fixes to tweak the style in whichever way they like. Or they could perform a bunch of AST transforms using Recast and then apply Prettier to format everything.
+
+### Style Editor Composition
+
+Because style editors are just objects, end-users can compose new style editors from existing ones. For example, if someone wanted to use an Airbnb style editor but tweak the indentation scheme a bit, they could create a new style editor like this:
+
+```js
+const airbnbStyle = require("eslint-plugin-airbnb").styles.default;
+
+module.exports = {
+    meta: {
+        name: "Modified Airbnb Style",
+        description: "A style editor based on Airbnb"
+    },
+    edits: [
+        ...airbnbStyle.edits,
+        require("./edits/indentation")
+    ]
+}
+```
+
+This style editor first applies the edits from the Airbnb style editor and then applies a custom indentation scheme. The possibilities for mixing and matching style editors are limitless, and all of it can be done without involving the ESLint team.
+
+### Order of Operations
+
+When ESLint is called with the `--fix` or `--fix-dry-run` flag, the style editor is applied to the source code after lint fixes are applied and before one final lint without fixes. The complete order of operations is as follows:
+
+1. Preprocess source code using configured processors (if any)
+1. Lint source code
```

Comment:
As a performance optimization, it should be possible to run only those rules that specify the `fixable` property in this step.